oracle/oracles/pd.go

```diff
@@ -241,28 +320,145 @@ func (o *pdOracle) getLastTSWithArrivalTS(txnScope string) (*lastTSO, bool) {
+// passed to this function as the requiredStaleness parameter.
+func (o *pdOracle) nextUpdateInterval(now time.Time, requiredStaleness time.Duration) time.Duration {
+	o.adaptiveUpdateIntervalState.mu.Lock()
+	defer o.adaptiveUpdateIntervalState.mu.Unlock()
+
+	configuredInterval := time.Duration(o.lastTSUpdateInterval.Load())
+	currentAdaptiveInterval := time.Duration(o.adaptiveLastTSUpdateInterval.Load())
+	if configuredInterval <= minAllowedAdaptiveUpdateTSInterval {
+		// If the user has configured a very short interval, we don't have any space to adjust it. Just use
+		// the user's configured value directly.
+		if currentAdaptiveInterval != configuredInterval {
+			o.adaptiveLastTSUpdateInterval.Store(int64(configuredInterval))
+		}
+		if o.adaptiveUpdateIntervalState.state != adaptiveUpdateTSIntervalStateUnadjustable {
+			logutil.Logger(context.Background()).Info("update low resolution ts interval is not being adaptive because the configured interval is too short",
+				zap.Duration("configuredInterval", configuredInterval),
+				zap.Stringer("state", o.adaptiveUpdateIntervalState.state),
+				zap.Stringer("newState", adaptiveUpdateTSIntervalStateUnadjustable))
+			o.adaptiveUpdateIntervalState.state = adaptiveUpdateTSIntervalStateUnadjustable
+		}
+		return configuredInterval
+	}
+
+	if requiredStaleness != 0 {
+		if requiredStaleness < currentAdaptiveInterval && currentAdaptiveInterval > minAllowedAdaptiveUpdateTSInterval {
+			// If we are calculating the interval because of a request that requires a shorter staleness, we shrink the
+			// update interval immediately to adapt to it.
+			// We shrink the update interval to a value slightly lower than the requested staleness to avoid potential
+			// frequent shrinking operations. But there's a lower bound to prevent loading ts too frequently.
+			prevAdaptiveInterval := currentAdaptiveInterval
+			currentAdaptiveInterval = max(requiredStaleness-adaptiveUpdateTSIntervalShrinkingPreserve, minAllowedAdaptiveUpdateTSInterval)
+			o.adaptiveLastTSUpdateInterval.Store(int64(currentAdaptiveInterval))
+			logutil.Logger(context.Background()).Info("shrink low resolution ts update interval immediately",
+				zap.Duration("requestedStaleness", requiredStaleness),
+				zap.Duration("prevAdaptiveUpdateInterval", prevAdaptiveInterval),
+				zap.Duration("newAdaptiveUpdateInterval", currentAdaptiveInterval),
+				zap.Stringer("prevState", o.adaptiveUpdateIntervalState.state),
+				zap.Stringer("newState", adaptiveUpdateTSIntervalStateAdapting))
+			o.adaptiveUpdateIntervalState.state = adaptiveUpdateTSIntervalStateAdapting
+		}
+		return currentAdaptiveInterval
+	}
+
+	lastReachDropThresholdTime := time.UnixMilli(o.adaptiveUpdateIntervalState.lastReachDropThresholdTime.Load())
+	if now.Sub(lastReachDropThresholdTime) < adaptiveUpdateTSIntervalDelayBeforeRecovering && currentAdaptiveInterval != configuredInterval {
```

Comment:
Should we check the `currentAdaptiveInterval != configuredInterval` first as `currentAdaptiveInterval` equals `configuredInterval` most of the time?